pxr/usd/usdGeom/validators.cpp

```diff
@@ -0,0 +1,136 @@
+//
+// Copyright 2024 Pixar
+//
+// Licensed under the terms set forth in the LICENSE.txt file available at
+// https://openusd.org/license.
+//
+#include "pxr/base/tf/registryManager.h"
+#include "pxr/base/tf/stringUtils.h"
+#include "pxr/base/tf/token.h"
+#include "pxr/usd/usd/prim.h"
+#include "pxr/usd/usd/schemaRegistry.h"
+#include "pxr/usd/usd/validationError.h"
+#include "pxr/usd/usd/validationRegistry.h"
+#include "pxr/usd/usd/validator.h"
+#include "pxr/usd/usdGeom/imageable.h"
+#include "pxr/usd/usdGeom/subset.h"
+#include "pxr/usd/usdGeom/validatorTokens.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+static
+UsdValidationErrorVector
+_SubsetFamilies(const UsdPrim& usdPrim)
+{
+    if (!(usdPrim && usdPrim.IsInFamily<UsdGeomImageable>(
+            UsdSchemaRegistry::VersionPolicy::All))) {
+        return {};
+    }
+
+    const UsdGeomImageable imageable(usdPrim);
+    if (!imageable) {
+        return {};
+    }
+
+    const TfToken::Set subsetFamilyNames =
+        UsdGeomSubset::GetAllGeomSubsetFamilyNames(imageable);
+
+    // Sort the family names so that they are in dictionary order, making
+    // the order in which they are validated more predictable.
+    TfTokenVector subsetFamilyNamesVec(
+        subsetFamilyNames.begin(), subsetFamilyNames.end());
+    std::sort(
+        subsetFamilyNamesVec.begin(), subsetFamilyNamesVec.end(),
+        TfDictionaryLessThan());
+
+    UsdValidationErrorVector errors;
+
+    for (const TfToken& subsetFamilyName : subsetFamilyNamesVec) {
+        const std::vector<UsdGeomSubset> familySubsets =
+            UsdGeomSubset::GetGeomSubsets(
+                imageable,
+                /* elementType = */ TfToken(),
+                /* familyName = */ subsetFamilyName);
+
+        // Determine the element type of the family by looking at the
```

Comment:
I wasn't completely certain about this. Is it possible/useful for the subsets of a given family to mix element types, or must they all have the same element type? I *think* the intention is the latter given what `ValidateFamily()` does here:
https://github.com/PixarAnimationStudios/OpenUSD/blob/5332d627cb28991e812acd67b8971e87367d9191/pxr/usd/usdGeom/subset.cpp#L831

But I don't actually think it's possible to end up in the code at the moment, since the `GetGeomSubsets()` call above it specifies an `elementType` which I don't think can be empty?

Reply:
The intention is for all the subsets in a family to have the same elementType!

The elementType in `GetGeomSubsets` can be empty, in which case all subsets in the family regardless of elementType will be returned. That being said, the `GetGeomSubsets()` call does pass in elementType, so great catch about that code never being hit! I'll update this code and the test for it separately.

Reply:
Great, thanks for clarifying @diyajoy!

Yeah, `GetGeomSubsets()` does allow for `elementType` to be empty, but the `_ValidateGeomType()` call at the top of `ValidateFamily()` basically ensures that it won't be, so our call to `GetGeomSubsets()` will be given a non-empty `elementType` and filter out all other element type subsets, preventing us from seeing whether there's a mix of element types.

Once we have that sorted out, I can add a test case for this in a follow-up PR to make sure we catch the case where a subset family contains subsets with differing element types and that we flag those as validation errors.

Thanks again!

Reply:
Actually if it's easy enough, would you mind including a fix for the `ValidateFamily` issue in this PR? It should just be passing in an empty `TfToken` for the `elementType` in the call to `GetGeomSubsets()`!

Reply:
A follow-up PR sounds great, thank you!

Since `elementType` will always be non-empty, I'm thinking that if we pass in an empty `TfToken` directly instead of the `elementType`, we'll get all subsets in the family regardless of their type.

Reply:
It was indeed easy enough, so I just went ahead and made the change. :)

The fix is exactly as you described in d9cc6f344ee3c537cf597bc7dc5648f4777b183b, and then I incorporated a `mixedElementTypes` subset family in the unit test in c3d111878f93754775c483504f34e8a91cbdbeb7.

Thanks!